File: core/src/main/java/io/squashql/query/dto/Period.java
package io.squashql.query.dto;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import io.squashql.query.field.Field;
import java.util.Set;

/**
 * TODO
 * MONTH_FROM_DATE,
 * QUARTER_FROM_DATE,
 * QUARTER_FROM_MONTH_YEAR,
 * SEMESTER_FROM_DATE,
 * YEAR_FROM_DATE
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS)
public interface Period {

  @JsonIgnore
  Set<Field> getFields();

  record Month(Field month, Field year) implements Period {

    @Override
    public Set<Field> getFields() {
      return Set.of(this.month, this.year);
    }
  }

  record Quarter(Field quarter, Field year) implements Period {

    @Override
    public Set<Field> getFields() {
      return Set.of(this.quarter, this.year);
    }
  }

  record Semester(Field semester, Field year) implements Period {

    @Override
    public Set<Field> getFields() {
      return Set.of(this.semester, this.year);
    }
  }

  record Year(Field year) implements Period {

    @Override
    public Set<Field> getFields() {
      return Set.of(this.year);
    }
  }
}

File: js/typescript-library/src/period.ts
import PACKAGE from "./package"
import {Field} from "./field"

export interface Period {
  readonly class: string,
}

export class Month implements Period {
  readonly class: string = PACKAGE + "dto.Period$Month"

  constructor(private month: Field, private year: Field) {
  }

  toJSON() {
    return {
      "@class": this.class,
      "month": this.month,
      "year": this.year,
    }
  }
}

export class Quarter implements Period {
  readonly class: string = PACKAGE + "dto.Period$Quarter"

  constructor(private quarter: Field, private year: Field) {
  }

  toJSON() {
    return {
      "@class": this.class,
      "quarter": this.quarter,
      "year": this.year,
    }
  }
}

export class Semester implements Period {
  readonly class: string = PACKAGE + "dto.Period$Semester"

  constructor(private semester: Field, private year: Field) {
  }

  toJSON() {
    return {
      "@class": this.class,
      "semester": this.semester,
      "year": this.year,
    }
  }
}

export class Year implements Period {
  readonly class: string = PACKAGE + "dto.Period$Year"

  constructor(private year: Field) {
  }

  toJSON() {
    return {
      "@class": this.class,
      "year": this.year,
    }
  }
}
